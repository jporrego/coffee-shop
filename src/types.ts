export interface Product {
  id: number;
  name: string;
  price: number;
  description: string;
  img: string;
}

export type StringTouple = [string, string];
