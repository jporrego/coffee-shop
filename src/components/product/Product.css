.product {
    position: relative;
    display: grid;
    grid-template-columns: 1fr;
    grid-template-rows: 55% 6rem 1fr;
    flex-direction: column;
    justify-items: center;
    padding: 1.5rem 1rem;
    width: 16rem;
    height: 24.5rem;
    border-radius: 2rem;
    background-color: var(--grey);
    box-shadow: 0px 2px 15px rgba(0, 0, 0, .0);
    overflow: hidden;
}

.product img {
    height: 100%;
    width: auto;
    transition: all 0.2s ease-out;
}

.product-name {
    align-self: end;
    text-align: center;
    white-space: wrap;
    max-height: 100%;
    max-width: 100%;
    font-weight: 700;
    padding: .5rem 0rem;
}

.product-price {
    align-self: end;
    margin-top: .5rem;
    font-size: 1.6rem;
    font-weight: 700;
    color: var(--brown-light);
}

.product .btn-round {
    position: absolute;
    right: 1rem;
    bottom: 1rem;
}

.product:hover img {
    transform: scale(1.075);
}