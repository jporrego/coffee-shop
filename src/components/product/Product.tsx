import React, { useState } from "react";
import "./Product.css";
import { Product as ProductInterface } from "../../types";

interface ProductProps {
  product: ProductInterface;
  onAddToCart: (id: number) => void;
  openProductModal: (id: number) => void;
}

const Product: React.FC<ProductProps> = ({
  product,
  onAddToCart,
  openProductModal,
}) => {
  const { id, name, price, img } = product;
  const [showMessage, setShowMessage] = useState(false);

  return (
    <div
      className="product"
      onClick={() => !showMessage && openProductModal(id)}
    >
      <img src={require("../../assets/img/product/" + img)} alt="" />
      <div className="product-name">{name}</div>
      <div className="product-price">${price}.00</div>
      <button
        className="btn-round"
        onClick={(e) => {
          e.stopPropagation();
          onAddToCart(id);
          setShowMessage(true);
          setTimeout(() => setShowMessage(false), 1500);
        }}
      >
        +
      </button>
      {showMessage && (
        <div className="product-added-mesage">
          <div>Added to cart</div>
        </div>
      )}
    </div>
  );
};

export default Product;
