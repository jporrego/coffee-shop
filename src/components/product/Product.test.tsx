import React from "react";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import Product from "./Product";
import { Product as ProductInterface } from "../../types";

// setup props
const product: ProductInterface = {
  id: 0,
  name: "Product",
  price: 10,
  img: "breville.png",
  description: "Product description",
};

const onAddToCart = jest.fn();
const openProductModal = jest.fn();

describe("product", () => {
  it("renders the product", () => {
    // since screen does not have the container property, we'll destructure render to obtain container for this test
    const { container } = render(
      <Product
        product={product}
        onAddToCart={onAddToCart}
        openProductModal={openProductModal}
      />
    );
    expect(container).toMatchSnapshot();
  });

  it("renders the correct product information", () => {
    render(
      <Product
        product={product}
        onAddToCart={onAddToCart}
        openProductModal={openProductModal}
      />
    );
    expect(screen.getByText("Product")).toBeInTheDocument();
    expect(screen.getByText(/10/)).toBeInTheDocument();
  });

  it("calls onAddToCart after button click", () => {
    render(
      <Product
        product={product}
        onAddToCart={onAddToCart}
        openProductModal={openProductModal}
      />
    );
    const button = screen.getByRole("button", { name: "+" });
    userEvent.click(button);
    expect(onAddToCart).toHaveBeenCalledTimes(1);
  });
});
