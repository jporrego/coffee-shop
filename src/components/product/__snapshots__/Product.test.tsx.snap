// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`product renders the product 1`] = `
<div>
  <div
    class="product"
  >
    <img
      alt=""
      src="breville.png"
    />
    <div
      class="product-name"
    >
      Product
    </div>
    <div
      class="product-price"
    >
      $
      10
      .00
    </div>
    <button
      class="btn-round"
    >
      +
    </button>
  </div>
</div>
`;
